src/components/CourseFilter/CourseFilter.tsx

```diff
@@ -0,0 +1,192 @@
+  FormControl,
+  ButtonGroup,
+} from '@mui/material';
+import { useTheme } from '@mui/material/styles';
+import { Key, useCallback } from 'react';
+import { CourseWithTagsAndStudentCount } from '@/lib/prisma/courses';
+import { usePathname, useSearchParams, useRouter } from 'next/navigation';
+import DatePicker from 'react-datepicker';
+import 'react-datepicker/dist/react-datepicker.css';
+
+type CourseFilterProps = {
+  initialCourses: CourseWithTagsAndStudentCount[];
+  initialTags: any;
+};
+
+export default function CourseFilter({
+  initialCourses,
+  initialTags,
+}: CourseFilterProps) {
+  const router = useRouter();
+  const pathname = usePathname();
+  const searchParams = useSearchParams();
+
+  const { palette } = useTheme();
+
+  const [startDate, setStartDate] = useState(null);
+  const [endDate, setEndDate] = useState(null);
+  const [courseName, setCourseName] = useState('');
+  const [selectedTags, setSelectedTags] = useState<string[]>([]);
+
+  const createQueryString = useCallback(
+    (name: string, value: string) => {
+      const params = new URLSearchParams(searchParams);
+      params.set(name, value);
+
+      return params.toString();
+    },
+    [searchParams]
+  );
+
+  useEffect(() => {
+    if (
+      searchParams.get('courseName') === null &&
+      searchParams.get('courseTag') === null &&
+      searchParams.get('courseDates') === null &&
+      searchParams.get('courseId') === null
+    ) {
+      handleClearSearch();
+    }
+  }, [searchParams]);
+
+  const handleNameChange = async (value: string | null) => {
+    const searchTerm: string | null = value;
+    setCourseName(searchTerm || '');
+    router.push(
+      pathname + '?' + createQueryString('courseName', searchTerm || '')
+    );
+  };
+
+  const handleTagChange = async (tagName: string) => {
+    let updatedTags = [];
+    if (selectedTags.includes(tagName)) {
+      updatedTags = selectedTags.filter((tag) => tag !== tagName);
+      setSelectedTags(updatedTags);
+    } else {
+      updatedTags = [...selectedTags, tagName];
+      setSelectedTags(updatedTags);
+    }
+    if (tagName === '') {
+      setSelectedTags([]);
+    }
+    const tagsAsaString = updatedTags.filter((tag) => tag).join(',');
+    router.push(pathname + '?' + createQueryString('courseTag', tagsAsaString));
+  };
+
+  const handleDateChange = async (range: [any, any]) => {
+    const [startDate, endDate] = range;
+    setStartDate(startDate);
+    setEndDate(endDate);
+    if (startDate === null && endDate === null) {
+      router.push(pathname + '?' + createQueryString('courseDates', ''));
+    } else {
+      router.push(
+        pathname +
+          '?' +
+          createQueryString('courseDates', startDate + '-' + endDate)
+      );
+    }
+  };
+
+  const handleClearSearch = async () => {
+    await handleNameChange(null);
+    await handleTagChange('');
+    await handleDateChange([null, null]);
+    router.replace(pathname);
+  };
+
+  const customStyles = {
```

Comment:
Could rename the styles variable to be connected to the use case. Won't have to search then where the style is applied to. F.ex "datePickerInputStyle" and so on